flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java

```diff
@@ -184,78 +185,106 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
     }
 
-    pendingManifestMap.clear();
+    pendingMap.clear();
 
-    // Delete the committed manifests and clear the committed data files from dataFilesPerCheckpoint.
-    for (ManifestFile manifestFile : manifestFiles) {
+    // Delete the committed manifests.
+    for (ManifestFile manifest : manifests) {
       try {
-        table.io().deleteFile(manifestFile.path());
+        table.io().deleteFile(manifest.path());
       } catch (Exception e) {
         // The flink manifests cleaning failure shouldn't abort the completed checkpoint.
         String details = MoreObjects.toStringHelper(this)
             .add("flinkJobId", newFlinkJobId)
             .add("checkpointId", checkpointId)
-            .add("manifestPath", manifestFile.path())
+            .add("manifestPath", manifest.path())
             .toString();
         LOG.warn("The iceberg transaction has been committed, but we failed to clean the temporary flink manifests: {}",
             details, e);
       }
     }
   }
 
-  private void replacePartitions(List<DataFile> dataFiles, String newFlinkJobId, long checkpointId) {
+  private void replacePartitions(NavigableMap<Long, WriteResult> pendingResults, String newFlinkJobId,
+                                 long checkpointId) {
+    // Partition overwrite does not support delete files.
+    int deleteFilesNum = pendingResults.values().stream().mapToInt(r -> r.deleteFiles().length).sum();
+    Preconditions.checkState(deleteFilesNum == 0, "Cannot overwrite partitions with delete files.");
+
+    // Commit the overwrite transaction.
     ReplacePartitions dynamicOverwrite = table.newReplacePartitions();
 
     int numFiles = 0;
-    for (DataFile file : dataFiles) {
-      numFiles += 1;
-      dynamicOverwrite.addFile(file);
+    for (WriteResult result : pendingResults.values()) {
+      numFiles += result.dataFiles().length;
+      Arrays.stream(result.dataFiles()).forEach(dynamicOverwrite::addFile);
     }
 
-    commitOperation(dynamicOverwrite, numFiles, "dynamic partition overwrite", newFlinkJobId, checkpointId);
+    commitOperation(dynamicOverwrite, numFiles, 0, "dynamic partition overwrite", newFlinkJobId, checkpointId);
   }
 
-  private void append(List<DataFile> dataFiles, String newFlinkJobId, long checkpointId) {
-    AppendFiles appendFiles = table.newAppend();
+  private void commitDeltaTxn(NavigableMap<Long, WriteResult> pendingResults, String newFlinkJobId, long checkpointId) {
+    int deleteFilesNum = pendingResults.values().stream().mapToInt(r -> r.deleteFiles().length).sum();
 
-    int numFiles = 0;
-    for (DataFile file : dataFiles) {
-      numFiles += 1;
-      appendFiles.appendFile(file);
-    }
+    if (deleteFilesNum == 0) {
+      // To be compatible with iceberg format V1.
+      AppendFiles appendFiles = table.newAppend();
 
-    commitOperation(appendFiles, numFiles, "append", newFlinkJobId, checkpointId);
+      int numFiles = 0;
+      for (WriteResult result : pendingResults.values()) {
```

Comment:
here we are merging data files potentially from multiple checkpoint cycles/manifests into a single manifest file.

 We are using this API from `AppendFiles` interface. 
```
  AppendFiles appendManifest(ManifestFile file);
```

When we had an extended outage and accumulated a few hundreds of transactions/manifests in Flink checkpoint, this help avoiding rewrite of many manifest files. @rdblue can probably explain it better than I do.

Reply:
It sounds like a separate improvement , so I created an issue for this , let's discuss there, https://github.com/apache/iceberg/issues/1959.

Reply:
> Maybe we can add a similar API in DeleteFiles interface?

We don't currently do this because we need delete entries to exist when we delete files. That way we can track when something was deleted and clean it up incrementally in `ExpireSnapshots`. If we did have a method like this, it would always rewrite the manifest with deletes, or would need to ensure that the manifest that is added contains only deletes, and these requirements are not very obvious. I think it is better to pass the deleted files through the existing methods.